### PR TITLE
[avro-cpp] Update to latest version (2022-11-07)

### PR DESCRIPTION
**Pull request by PavelKisliak**
**Update 'avro-cpp' port to latest version 1.11.2 pre-release**

- #### What does your PR fix?
  - Update avro-cpp to new version
  - Remove setting DEBUG option 'AVRO_ADD_PROTECTOR_FLAGS' (leading to binary incompatibility and segfault crashes)
  - Add feature for build tests
  - Remove not used dependencies
  - Update deprecated VCPKG functions
  - Add 'usage' file

- #### Which triplets are supported/not supported? Have you updated the [CI baseline](https://github.com/microsoft/vcpkg/blob/master/scripts/ci.baseline.txt)?
  `All (only with shared linkage)`, `No`

- #### Does your PR follow the [maintainer guide](https://github.com/microsoft/vcpkg/blob/master/docs/maintainers/maintainer-guide.md)?
  `Yes`

- #### If you have added/updated a port: Have you run `./vcpkg x-add-version --all` and committed the result?
  `Yes`

### CODE REVIEW

File: ports/avro-cpp/vcpkg.json [hunk 1]
```diff
@@ -1,32 +1,40 @@
 {
   "name": "avro-cpp",
-  "version-date": "2021-06-01",
+  "version": "1.11.1",
   "port-version": 1,
   "description": "Apache Avro is a data serialization system",
   "homepage": "https://github.com/apache/avro",
   "supports": "!(windows & static)",
   "dependencies": [
+    "boost-algorithm",
```
Comment (MonicaLiu0311) [inline ports/avro-cpp/vcpkg.json:9]:
```suggestion
  "homepage": "https://github.com/apache/avro",
  "license": "Apache-2.0",
  "supports": "!(windows & static)",
  "dependencies": [
    "boost-algorithm",
```

Comment (MonicaLiu0311) [reply ports/avro-cpp/vcpkg.json:9]: A `license` is required.

File: ports/avro-cpp/portfile.cmake [hunk 2]
```diff
@@ -3,34 +3,31 @@ vcpkg_check_linkage(ONLY_DYNAMIC_LIBRARY)
 vcpkg_from_github(
     OUT_SOURCE_PATH SOURCE_PATH
     REPO apache/avro
-    REF 2ab8fa85d05f04387bd5d63b10ad1c8fd2243616
-    SHA512 fd21f0919b0e5e884bdf4d66c4d5ba056f04c426b309ec0b5ab26642a5f6b00d46f4dd965431b10130bc5f0d81699e2195780e90e127f63049ee5763403ef7c8
-    HEAD_REF master
```
Comment (MonicaLiu0311) [inline ports/avro-cpp/portfile.cmake:8]: After checking, I found that the latest upstream version is `release-1.11.1`, and the `master` is the main branch. Why delete it?

Comment (PavelKisliak) [reply ports/avro-cpp/portfile.cmake:8]: Agree, it's my mistake, thank you!

File: ports/avro-cpp/portfile.cmake [hunk 3]
```diff
@@ -3,34 +3,31 @@ vcpkg_check_linkage(ONLY_DYNAMIC_LIBRARY)
...
-    PREFER_NINJA
-    DISABLE_PARALLEL_CONFIGURE
+vcpkg_cmake_configure(
+    SOURCE_PATH "${SOURCE_PATH}/lang/c++"
     OPTIONS
-        -DBUILD_TESTING=OFF
         ${FEATURE_OPTIONS}
-    OPTIONS_DEBUG
-        -DAVRO_ADD_PROTECTOR_FLAGS=1
 )
```
Comment (MonicaLiu0311) [inline ports/avro-cpp/portfile.cmake:24]: `vcpkg` does not recommend building `tests`, and an error occurred when testing the feature. For details, see:

`./vcpkg install avro-cpp[test]`

```
-- Configuring x86-windows
-- Building x86-windows-dbg
CMake Error at scripts/cmake/vcpkg_execute_build_process.cmake:131 (message):
    Command failed: "C:/Program Files/CMake/bin/cmake.exe" --build . --config Debug --target install -- -v -j9
    Working Directory: E:/src/27825/buildtrees/avro-cpp/x86-windows-dbg
    See logs for more information:
      E:\src\27825\buildtrees\avro-cpp\install-x86-windows-dbg-out.log

Call Stack (most recent call first):
  scripts/cmake/vcpkg_build_cmake.cmake:78 (vcpkg_execute_build_process)
  scripts/cmake/vcpkg_install_cmake.cmake:18 (vcpkg_build_cmake)
  ports/avro-cpp/portfile.cmake:27 (vcpkg_install_cmake)
  scripts/ports.cmake:147 (include)

```
[install-x86-windows-dbg-out.log](https://github.com/microsoft/vcpkg/files/10018850/install-x86-windows-dbg-out.log)

File: ports/avro-cpp/vcpkg.json [hunk 4]
```diff
@@ -1,32 +1,41 @@
...
       "description": "Support Snappy for compression",
       "dependencies": [
         "snappy"
       ]
+    },
+    "test": {
+      "description": "Build tests",
+      "dependencies": [
+        "boost-test"
+      ]
```
Comment (MonicaLiu0311) [inline ports/avro-cpp/vcpkg.json:38]: An error occurred when testing the feature. For details, see:

`./vcpkg install avro-cpp[test]`

```
-- Configuring x86-windows
-- Building x86-windows-dbg
CMake Error at scripts/cmake/vcpkg_execute_build_process.cmake:131 (message):
    Command failed: "C:/Program Files/CMake/bin/cmake.exe" --build . --config Debug --target install -- -v -j9
    Working Directory: E:/src/27825/buildtrees/avro-cpp/x86-windows-dbg
    See logs for more information:
      E:\src\27825\buildtrees\avro-cpp\install-x86-windows-dbg-out.log

Call Stack (most recent call first):
  scripts/cmake/vcpkg_build_cmake.cmake:78 (vcpkg_execute_build_process)
  scripts/cmake/vcpkg_install_cmake.cmake:18 (vcpkg_build_cmake)
  ports/avro-cpp/portfile.cmake:27 (vcpkg_install_cmake)
  scripts/ports.cmake:147 (include)
```
[install-x86-windows-dbg-out.log](https://github.com/microsoft/vcpkg/files/10019004/install-x86-windows-dbg-out.log)

Comment (PavelKisliak) [reply ports/avro-cpp/vcpkg.json]: Fixed, thank you!

**Comment by PavelKisliak** [comment]
Just wondering why, there is still warning 'You have modified or added at least one portfile where deprecated functions are used.', I don't see any more deprecated functions from the list.

**Comment by MonicaLiu0311** [comment]
> Just wondering why there is still warning 'You have modified or added at least one portfile where deprecated functions are used.', I don't see any more deprecated functions from the list.

It may be this copyright function, and it will be changed next time.
`file(INSTALL "${SOURCE_PATH}/lang/c++/LICENSE" DESTINATION "${CURRENT_PACKAGES_DIR}/share/${PORT}" RENAME copyright)`
New functions:
`vcpkg_install_copyright(FILE_LIST "${SOURCE_PATH}/lang/c++/LICENSE")`

**Comment by MonicaLiu0311** [comment]
All features are tested successfully in the following triplet:
x86-windows
x64-windows

**Review by MonicaLiu0311 — DISMISSED** [review]

**Comment by dg0yt** [comment]
> may be this copyright function, and it will be changed next time.
> `file(INSTALL "${SOURCE_PATH}/lang/c++/LICENSE" DESTINATION "${CURRENT_PACKAGES_DIR}/share/${PORT}" RENAME copyright)`
> New functions:
> `vcpkg_install_copyright(FILE_LIST "${SOURCE_PATH}/lang/c++/LICENSE")`

No, this is not the reason.

File: ports/avro-cpp/portfile.cmake [hunk 5]
```diff
@@ -3,34 +3,31 @@ vcpkg_check_linkage(ONLY_DYNAMIC_LIBRARY)
...
 
 vcpkg_install_cmake(ADD_BIN_TO_PATH)
```
Comment (dg0yt) [inline ports/avro-cpp/portfile.cmake:26]: With `vcpkg_cmake_configure`, you must use `vcpkg_cmake_install`. This is the trigger for the deprecation notice.

File: ports/avro-cpp/usage [hunk 6]
```diff
@@ -0,0 +1,7 @@
+The package avro-cpp can be used from CMake via:
+
+    find_path(AVROCPP_INCLUDE_DIRS avro/Encoder.hh)
+    find_library(AVROCPP_LIBRARIES avrocpp REQUIRED)
+
+    target_include_directories(main PRIVATE ${AVROCPP_INCLUDE_DIRS})
+    target_link_libraries(main PRIVATE ${AVROCPP_LIBRARIES})
```
Comment (dg0yt) [inline ports/avro-cpp/usage:7]: IMO you should use singular variable names with `find_path` and `find_library`, and add quotes when using them in expressions or arguments.

In addition, I'm concerned that a single `find_library` will not work well with multi-config generators (in particular with Visual Studio).

**Review by dg0yt — CHANGES_REQUESTED** [review]

File: ports/avro-cpp/portfile.cmake (hunk 5 above)
Comment (MonicaLiu0311) [reply ports/avro-cpp/portfile.cmake:26]: You are right, it was my carelessness.

File: ports/avro-cpp/usage (hunk 6 above)
Comment (PavelKisliak) [reply ports/avro-cpp/usage]: Thanks, fixed!

File: ports/avro-cpp/portfile.cmake (hunk 5 above)
Comment (PavelKisliak) [reply ports/avro-cpp/portfile.cmake]: Thanks @dg0yt, fixed!

File: ports/avro-cpp/vcpkg.json [hunk 7]
```diff
@@ -1,32 +1,44 @@
...
+    "zlib"
   ],
   "features": {
     "snappy": {
       "description": "Support Snappy for compression",
       "dependencies": [
         "snappy"
       ]
+    },
+    "test": {
```
Comment (BillyONeal) [inline ports/avro-cpp/vcpkg.json:33]: I don't think this feature should exist: vcpkg should not be building tests.

Comment (PavelKisliak) [reply ports/avro-cpp/vcpkg.json:33]: I was just following official [maintainer-guide](https://github.com/microsoft/vcpkg/blob/master/docs/maintainers/maintainer-guide.md#do-not-build-testsdocsexamples-by-default), this is an optinal feature which is not build by default. I see that there are already existing a lot of ports with such feature.

Comment (dg0yt) [reply ports/avro-cpp/vcpkg.json:33]: @BillyONeal IMO the recommendation needs to be "optional feature", not "this feature shouldn't exist". Given a broad usage of vcpkg ports beyond platforms tested by upstream or vcpkg CI, the lack of an easy way to run tests *on demand* is a disservice for users.

**Review by MonicaLiu0311 — APPROVED** [review]

**Review by MonicaLiu0311 — APPROVED** [review]

**Comment by JavierMatosD** [comment]
Thank you

**Pull request merged**